### ISSUE TITLE
Missing babel edge case compiler error

### ISSUE THREAD

**Issue by chrbala** [eslint/eslint#5082]
**What version are you using?**
1.10.3 via NPM on sublimetext

**What happened?**

I get a compiler error when I do 
``Content || () => <div />``
but ESLint doesn’t catch it. I have to do 
``Content || (() => <div />)``
in order to get the compiler to work

**Comment by nzakas** [comment]
We need a lot more information to look into this.

1. What do you mean by compiler error? Are you talking about a Babel error?
2. Please share the full source code (your code examples causes a parsing error in ESLint)
3. What configuration are you using for ESLint?
4. Please paste the exact output from ESLint

**Comment by chrbala** [comment]
Ah, what I meant was that Babel was not compiling with ``Content || () => <div />`` . ESLint was working without error, but It'd make for a nicer experience if ESLint caught the error.

I don't really have a particular output for ESLint on this issue, since the problem is a false negative.

ESLint config below.

```
"parser": "babel-eslint",
"rules": {
  "strict": 0,
  "semi": [
    2,
    "never"
  ],
  "comma-dangle": [
    2,
    "never"
  ],
  "no-dupe-keys": 2,
  "no-func-assign": 2,
  "no-sparse-arrays": 2,
  "no-unreachable": 2,
  "use-isnan": 2,
  "valid-typeof": 2,
  "no-cond-assign": 2,
  "no-duplicate-case": 2,
  "no-unexpected-multiline": 2,
  "accessor-pairs": 2,
  "block-scoped-var": 2,
  "curly": [
    2,
    "multi"
  ],
  "dot-notation": 2,
  "no-caller": 2,
  "no-case-declarations": 2,
  "no-else-return": 2,
  "no-empty-pattern": 2,
  "no-eq-null": 2,
  "no-eval": 2,
  "no-extend-native": [
    2,
    {
      "exceptions": [
        "Object"
      ]
    }
  ],
  "no-extra-bind": 2,
  "no-fallthrough": 2,
  "no-implied-eval": 2,
  "no-loop-func": 2,
  "no-multi-spaces": 2,
  "no-multi-str": 2,
  "no-native-reassign": 2,
  "no-new-func": 2,
  "no-new-wrappers": 2,
  "no-new": 2,
  "no-octal-escape": 2,
  "no-octal": 2,
  "no-proto": 2,
  "no-redeclare": 2,
  "no-return-assign": 2,
  "no-script-url": 2,
  "no-self-compare": 2,
  "no-unused-expressions": [
    2,
    {
      "allowShortCircuit": true,
      "allowTernary": true
    }
  ],
  "no-useless-call": 2,
  "no-useless-concat": 2,
  "no-void": 2,
  "no-warning-comments": [
    1,
    {
      "terms": [
        "todo"
      ],
      "location": "anywhere"
    }
  ],
  "no-with": 2,
  "radix": 2,
  "wrap-iife": [
    2,
    "inside"
  ],
  "yoda": 2,
  "no-catch-shadow": 2,
  "no-delete-var": 2,
  "no-shadow-restricted-names": 2,
  "no-shadow": [
    2,
    {
      "builtinGlobals": true
    }
  ],
  "no-undef-init": 2,
  "no-use-before-define": 2,
  "brace-style": 2,
  "camelcase": 2,
  "comma-spacing": [
    2,
    {
      "before": false,
      "after": true
    }
  ],
  "comma-style": [
    2,
    "last"
  ],
  "computed-property-spacing": [
    2,
    "never"
  ],
  "id-length": [
    2,
    {
      "min": 2,
      "max": 25,
      "exceptions": [
        "e",
        "i",
        "j",
        "m",
        "$"
      ]
    }
  ],
  "key-spacing": [
    2,
    {
      "beforeColon": false,
      "afterColon": true,
      "mode": "minimum"
    }
  ],
  "linebreak-style": [
    2,
    "unix"
  ],
  "max-depth": 2,
  "max-len": [
    1,
    80,
    2
  ],
  "max-nested-callbacks": [
    2,
    3
  ],
  "max-params": [
    2,
    3
  ],
  "new-parens": 2,
  "no-array-constructor": 2,
  "no-bitwise": 2,
  "no-continue": 2,
  "no-lonely-if": 2,
  "no-multiple-empty-lines": [
    2,
    {
      "max": 1
    }
  ],
  "no-nested-ternary": 2,
  "no-new-object": 2,
  "no-unneeded-ternary": 2,
  "one-var": [
    2,
    "never"
  ],
  "operator-assignment": [
    2,
    "always"
  ],
  "operator-linebreak": [
    2,
    "none",
    {
      "overrides": {
        "?": "before",
        ":": "before",
        "&&": "before",
        "||": "before"
      }
    }
  ],
  "quote-props": [
    2,
    "as-needed"
  ],
  "space-after-keywords": 2,
  "space-before-blocks": 2,
  "space-before-function-paren": [
    2,
    "never"
  ],
  "space-before-keywords": [
    2,
    "always"
  ],
  "space-infix-ops": 2,
  "space-return-throw-case": 2,
  "space-unary-ops": [
    1,
    {
      "words": true,
      "nonwords": false
    }
  ],
  "spaced-comment": [
    2,
    "always",
    {
      "exceptions": [
        "*"
      ]
    }
  ],
  "arrow-body-style": [
    2,
    "as-needed"
  ],
  "arrow-parens": [
    2,
    "as-needed"
  ],
  "arrow-spacing": 2,
  "constructor-super": 2,
  "no-class-assign": 2,
  "no-const-assign": 2,
  "no-dupe-class-members": 2,
  "no-this-before-super": 2,
  "object-shorthand": 2,
  "prefer-arrow-callback": 2,
  "prefer-spread": 2,
  "dot-location": [
    2,
    "property"
  ]
}
```

**Comment by nzakas** [comment]
Can you try without babel-eslint?

**Comment by chrbala** [comment]
Without babel-eslint, it creates an error because I'm using babel features - unless I'm missing a way of parsing Babel without babel-eslint.

This is a complete example of the code I'm trying to test:
```
var hello = false
var Component = hello || () => <div />
```

Without babel-eslint, it has a parsing error with the ``() =>`` arrow syntax.

**Comment by hzoo** [comment]
@chrbala You can use ecmaFeatures http://eslint.org/docs/user-guide/configuring#specifying-language-options with jsx/arrow functions

**Comment by chrbala** [comment]
Having tried again with
```
"ecmaFeatures": {
  "arrowFunctions": true
}
```
and also removing the JSX, ESLint gives me a parsing error on:
```
var hello = false
var Component = hello || () => true
``` 

and it gives no error on
```
var hello = false
var Component = () => true
``` 
So it seems like it's a babel-eslint issue.

**Comment by michaelficarra** [comment]
Arrows have the precedence of assignment expressions. That they don't in babel is a known bug. Eslint is correctly giving you a parsing error.

**Comment by jmm** [comment]
Arrow parsing precedence was fixed a long time ago in Babel (http://astexplorer.net/#/NL0Fd3mEgg), no?

@chrbala What version of `babel-eslint` are you using?

**Comment by nzakas** [comment]
@chrbala I'd suggest following up with babel-eslint for this.
